sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConsumeValueTaskCorrectly.cs

```diff
@@ -0,0 +1,225 @@
+
+            public ConsumeValueTaskWalker(SemanticModel semanticModel)
+            {
+                this.semanticModel = semanticModel;
+                SymbolUsages = new Dictionary<ISymbol, IList<SyntaxNode>>();
+                ConsumedButNotCompleted = new List<SyntaxNode>();
+            }
+
+            /**
+             * Check if 'await' is done on a 'ValueTask'
+             */
+            public override void VisitAwaitExpression(AwaitExpressionSyntax node)
+            {
+                if (node.Expression is IdentifierNameSyntax identifierName &&
+                    this.semanticModel.GetSymbolInfo(identifierName).Symbol is ISymbol symbol &&
+                    symbol.GetSymbolType().OriginalDefinition.IsAny(ValueTaskTypes))
+                {
+                    AddToSymbolUsages(symbol, identifierName);
+                }
+
+                base.VisitAwaitExpression(node);
+            }
+
+            /**
+             * Check if it's the wanted method on a ValueTask
+             * - we treat AsTask() like await - always add it to the list
+             * - for GetAwaiter().GetResult() - ignore the call if it's called inside an 'if (valueTask.IsCompletedSuccessfully)'
+             */
+            public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+            {
+                if (node.Expression is MemberAccessExpressionSyntax memberAccess)
+                {
+                    if (node.IsMethodInvocation(ValueTaskTypes, "AsTask", semanticModel) &&
+                        GetLeftMostIdentifier(memberAccess) is IdentifierNameSyntax identifier1)
+                    {
+                        AddToSymbolUsages(this.semanticModel.GetSymbolInfo(identifier1).Symbol, identifier1);
+                    }
+
+                    if (node.IsMethodInvocation(ValueTaskTypes, "GetResult", semanticModel) &&
+                        GetLeftMostIdentifier(memberAccess) is IdentifierNameSyntax identifier2 &&
+                        this.semanticModel.GetSymbolInfo(identifier2).Symbol is ISymbol symbol2 &&
+                        !IsCompletedSuccessfullyHasBeenCheckedInParent(symbol2, identifier2))
+                    {
+                        AddToSymbolUsages(symbol2, identifier2);
+                        ConsumedButNotCompleted.Add(identifier2);
+                    }
+                }
+
+                base.VisitInvocationExpression(node);
+            }
+
+            /**
+             * Check if ".Result" is accessed on a 'ValueTask'
+             * - ignore the call if it's called inside an 'if (valueTask.IsCompletedSuccessfully)'
+             */
+            public override void VisitMemberAccessExpression(MemberAccessExpressionSyntax node)
+            {
+                if (node.IsPropertyInvocation(ValueTaskTypes, "Result", semanticModel) &&
+                    GetLeftMostIdentifier(node) is IdentifierNameSyntax identifierName &&
```

Comment:
I need to get the identifier so that I can get the symbol of the variable, and keep a list of its usages (to see if it's used twice).